ecommerce/cybersource_api.py

```diff
@@ -0,0 +1,99 @@
+from ecommerce.constants import TRANSACTION_TYPE_REFUND, REFUND_SUCCESS_STATES
+from mitol.payment_gateway.api import (
+    Refund as GatewayRefund,
+    PaymentGateway,
+)
+from mitol.payment_gateway.exceptions import RefundDuplicateException
+
+log = logging.getLogger(__name__)
+
+
+def refund_order(*, order_id: int, **kwargs):
+    """
+    A function that performs refund for a given order id
+
+    Args:
+       order_id (int): Id of the order which is being refunded
+       kwargs (dict): Dictionary of the other attributes that are passed e.g. refund amount, refund reason, unenroll
+       If no refund_amount is provided it will use refund amount from Transaction obj
+       unenroll will never be performed if the refund fails
+
+    Returns:
+        tuple(bool, bool):
+        First bool identifies if the refund operation was a success or failure
+        Second bool identifies if unrenroll was success or failure
+    """
+    refund_amount = kwargs.get("refund_amount")
+    refund_reason = kwargs.get("refund_reason", "")
+    unenroll = kwargs.get("unenroll", False)
+
+    with transaction.atomic():
+        # Get order details for the given order_id (Optimally, This should be a fulfilled order)
+        order = FulfilledOrder.objects.select_for_update().get(id=order_id)
+
+        if order.state != Order.STATE.FULFILLED:
+            log.debug(f"Order with order_id {order_id} is not in fulfilled state.")
+            return False, False
+
+        # Extract transaction_id from order CyberSource json
+        order_recent_transaction = Transaction.objects.filter(order=order_id).first()
+
+        if not order_recent_transaction:
+            log.error(
+                f"There is no associated transaction against order_id {order_id}."
+            )
+            return False, False
+
+        transaction_dict = json.loads(json.dumps(order_recent_transaction.data))
+
+        # This should break if there is no transaction_id in transaction_data because that's needed for refund API
+        transaction_id = transaction_dict["transaction_id"]
+
+        order_amount = refund_amount or transaction_dict.get("req_amount")
+        order_currency = transaction_dict.get("req_currency")
+
+        refund_gateway_request = GatewayRefund(
+            transaction_id=transaction_id,
+            refund_amount=order_amount,
+            refund_currency=order_currency,
+        )
+
+        try:
+            response = PaymentGateway.start_refund(
+                ECOMMERCE_DEFAULT_PAYMENT_GATEWAY,
+                refund_gateway_request,
+            )
+            if response.state in REFUND_SUCCESS_STATES:
+                # There is no default reason for response of a Refund, so we append the response json with the reason
+                # provided by the caller
+                if refund_reason:
+                    response.response_data["reason"] = refund_reason
+
+                unenrolled = order.refund(
```

Comment:
We can have two types of implementations here:

1- We keep it less coupled and perform only API calls and return the results and let the caller update the other things e.g. making appropriate changes in database for marking the order as `Refunded`. But in this case the callers from different places will have to perform the database level refunds at multiple places/.

2- We keep it the way it is right now, More coupled but it will make almost all the necessary changes upon the API call e.g. Marking order refunded, unenrolling based on the passed flag.

@rhysyngsun Let me know what you think.

Reply:
Hmm, yeah, this is a bit tricky here. I think we need to decouple this because if things fail a certain way we could end up having made the api call to refund the learner via cybersource but then the `Order` save fails and rolls back so we end up in a bad state. And `Order.refund()` is doing way too much - it shouldn't be doing the unenrollments. I'd move that to a separate function and then reorder the logic in `refund_order` to this:

- Start a DB transaction, within that transaction:
  - Get the order and call `order.refund()`. This will ensure that the DB has the correct state before we make the call to Cybersource. This way if the DB write fails, we don't attempt the API call and if the API call fails the DB update rolls back
  - Make the call to the Cybersource refund API
- Exit the transaction so it commits
- Trigger the unenrollments, this might be best as a celery task because that could also fail temporarily, but at this point the most important work of issuing the refund has been completed so it's not absolutely necessary that this happen immediately.

I hope that answers your question?

Reply:
This does, It's a great suggestion. I had a thought on how the end-user behavior would work in the case of Refund through Admin. Running the unenrollments in a Celery task will have to change that behavior a bit. Currently, when a user clicks the refund what happens is Order refunds in DB, CyberSource API called, and lastly unenrollment, and when all that is done the User sees that the order got refunded.

After using the celery task, should we care for the unenrollment status or just return the user with success before the while the unenrollment celery task runs?

Reply:
@rhysyngsun an update on this part:

>Get the order and call order.refund(). This will ensure that the DB has the correct state before we make the call to Cybersource. This way if the DB write fails, we don't attempt the API call and if the API call fails the DB update rolls back

Ideally, when refunding the order we also set the acuired response json from CyberSoure refund API in transaction data which isn't available before the API call, So, i tweaked it in a way that it will refund the order, create the transaction and when we have api response it will update that transaction with response data.

>Trigger the unenrollments, this might be best as a celery task because that could also fail temporarily, but at this point the most important work of issuing the refund has been completed so it's not absolutely necessary that this happen immediately.

Updates, so in this case when an order is refunded from Django Admin, it refunds and starts the celery task. I updated the success message on admin to something like `Order <id> refunded and unenrollment is in progress.`.

Let me know if above seems feasible to you?